cpp/src/io/parquet/experimental/dictionary_page_filter.cu

```diff
@@ -46,11 +47,1480 @@ using parquet::detail::ColumnChunkDesc;
 using parquet::detail::decode_error;
 using parquet::detail::PageInfo;
 
+namespace {
+
+namespace cg = cooperative_groups;
+
+// Decode kernel parameters
+auto constexpr int96_size        = 12;   ///< Size of INT96 physical type
+auto constexpr decode_block_size = 128;  ///< Must be a multiple of warp_size
+/// Maximum number of literals to evaluate while decoding column dictionaries
+auto constexpr max_inline_literals = 2;
+auto constexpr fp_error_tolerance  = 1e-5;  ///< Floating point error tolerance
+
+// cuCollections static set parameters
+using key_type = cudf::size_type;                  ///< Using column indices (size_type) as set keys
+auto constexpr empty_key_sentinel = key_type{-1};  ///< We will never encounter a -1 index.
+auto constexpr set_cg_size        = 1;             ///< Cooperative group size for cuco::static_set
+auto constexpr bucket_size        = 1;             ///< Number of buckets per set slot
+auto constexpr occupancy_factor = 70;  ///< cuCollections suggests targeting a 70% occupancy factor
+uint32_t constexpr hasher_seed  = 0;   ///< default seed for the hashers
+
+/// Hash function for cuco::static_set_ref
+template <typename T>
+using hasher_type = cudf::hashing::detail::MurmurHash3_x86_32<T>;
+
+/// cuco::static_set_ref storage type
+using storage_type     = cuco::bucket_storage<key_type,
+                                              bucket_size,
+                                              cuco::extent<std::size_t>,
+                                              cudf::detail::cuco_allocator<char>>;
+using storage_ref_type = typename storage_type::ref_type;
+using bucket_type      = typename storage_type::bucket_type;
+
+/**
+ * @brief Helper function to check if two values are equal
+ *
+ * @tparam T Underlying data type of the cudf column
+ * @param lhs Left hand side value
+ * @param rhs Right hand side value
+ *
+ * @return Boolean indicating if the two values are equal
+ */
+template <typename T>
+__host__ __device__ __forceinline__ constexpr bool are_values_equal(T const& lhs,
+                                                                    T const& rhs) noexcept
+{
+  if constexpr (cudf::is_floating_point<T>()) {
+    return cuda::std::abs(lhs - rhs) < fp_error_tolerance;
```

Comment:
I know that we don't guarantee floating point errors from arithmetic in libcudf but we do need it here so that we may not accidentally prune a row group if a literal value did exist in a column chunk dictionary. Thoughts @PointKernel @vuule @bdice?

Reply:
For context, here we are evaluating if a parquet row group can be discarded if we have a filter expression of the form `col == literal`, by looking up the `literal` in each row group's dictionary and discarding it if it's not present.

Reply:
This is a little strange to me. I think we should use exact equality here. I think a user can be expected to do the same math for generating the value as for looking it up. Maybe that's unreasonable, I don't know.

Reply:
My initial though was: We are okay if the error tolerance results in a false positive membership as we apply a `cudf::compute_column` on the output table in the reader. Otherwise, we may prune a row group that was needed. Thinking more, even if we use this tolerance and force-keep the row groups, the rows of interest may still be pruned in the `cudf::compute_column`.

Reply:
Right. If the `cudf::compute_column` will use exact equality, we should apply the same rules here. I don't think we can avoid the fact that binary/exact equality of floats is required.

Reply:
>I think a user can be expected to do the same math for generating the value as for looking it up.
I think that's a reasonable expectation. Let me see what other readers that support dictionary based row group pruning do as well.

Reply:
>Right. If the cudf::compute_column will use exact equality, we should apply the same rules here. I don't think we can avoid the fact that binary/exact equality of floats is required.

Makes sense. Thanks @bdice